packages/aws-cdk/lib/logging.ts

```diff
@@ -114,99 +53,192 @@ export async function withCorkedLogging<T>(block: () => Promise<T>): Promise<T>
+  /**
+   * The log level to use
+   */
+  level: IoMessageLevel;
+  /**
+   * The message to log
+   */
+  message: string;
+  /**
+   * Whether to force stdout
+   * @default false
+   */
+  forceStdout?: boolean;
+  /**
+   * The message code
+   * @default TOOLKIT_2000, TOOLKIT_1000, or TOOLKIT_0000 depending on log level
+   */
+  code?: string;
+}
+
 /**
- * Core logging function that handles all log output.
+ * Core logging function that utilizes the CLI IO host to write messages to the console.
  * @param entry - LogEntry object or log level
  * @param fmt - Format string (when using with log level)
  * @param args - Format arguments (when using with log level)
  */
-export function log(entry: LogEntry): void;
-export function log(level: LogLevel, fmt: string, ...args: unknown[]): void;
-export function log(levelOrEntry: LogLevel | LogEntry, fmt?: string, ...args: unknown[]): void {
-  // Normalize input
-  const entry: LogEntry = typeof levelOrEntry === 'string'
-    ? { level: levelOrEntry as LogLevel, message: util.format(fmt!, ...args) }
-    : levelOrEntry;
-
+export function log(options: LogOptions) {
   // Check if we should log this level
-  if (levelPriority[entry.level] > levelPriority[currentLogLevel]) {
+  if (levelPriority[options.level] > levelPriority[currentIoMessageThreshold]) {
     return;
   }
 
-  // Format the message
-  let finalMessage = entry.message;
-
-  // Add timestamp first if requested
-  if (entry.timestamp) {
-    finalMessage = `[${formatTime(new Date())}] ${finalMessage}`;
-  }
-
-  // Add prefix AFTER timestamp
-  if (entry.prefix) {
-    finalMessage = `${entry.prefix} ${finalMessage}`;
+  // Determine if a provided message code is valid, otherwise if no message code is provided, defaults
+  // to the appropriate generic TOOLKIT code. 0000 for error, 1000 for warning, and 2000 for info and above.
+  if (options.code) {
+    if (!validateMessageCode(options.code)) {
+      throw new Error(`Invalid message code: ${options.code}`);
+    }
+  } else {
+    options.code = `TOOLKIT_${Math.min(levelPriority[options.level] - 1, 2)}000`;
   }
 
-  // Apply custom style if provided, otherwise use level-based style
-  const style = entry.style || styleMap[entry.level];
-  finalMessage = style(finalMessage);
-
   // Get appropriate stream - pass through forceStdout flag
-  const stream = getStream(entry.level, entry.forceStdout);
+  const stream = CliIoHost.getStream(options.level, options.forceStdout ?? false);
 
   // Handle corking
   if (CORK_COUNTER > 0) {
-    logBuffer.push([stream, finalMessage]);
+    logBuffer.push([stream, options.message]);
     return;
   }
 
+  const ioMessage: IoMessage = {
+    level: options.level,
+    message: options.message,
+    forceStdout: options.forceStdout,
+    time: new Date(),
+    action: CliIoHost.currentAction ?? 'none',
+    code: options.code,
+  };
+
   // Write to stream
-  stream.write(finalMessage + '\n');
+  void CliIoHost.getIoHost().notify(ioMessage);
```

Comment:
I fully welcome someone telling me we shouldn't do this for some reason (if only that I think it might create a corking race condition), but if we don't do this we gotta do `await error('asdfasdf')` instead of just `('asdfasdf')` since notify is an async function per [the rfc](https://github.com/aws/aws-cdk-rfcs/pull/654/files#diff-12642c528959289d826aebbb25c73a35b1d46bfb01b7156776daf15db0965efbR160)